stor/dx.py

```
@@ -0,0 +1,513 @@
+                return True
+            except NotFoundError:
+                return False
+        return False
+
+    def stat(self):
+        if not self.__stat:
+
+            if self.canonical_resource:
+                self.__stat = dxpy.DXFile(dxid=self.canonical_resource,
+                                          project=self.canonical_project).describe()
+            else:
+                self.__stat = dxpy.DXProject(dxid=self.canonical_project).describe()
+        return self.__stat
+
+
+class DXVirtualPath(DXPath):
+    """Class Handler for DXPath of form 'dx://project-{ID}:/a/b/c' or 'dx://a/b/c'"""
+
+    @property
+    def virtual_project(self):
+        if utils.is_valid_dxid(self.project, 'project'):
+            return dxpy.DXProject(dxid=self.project).name
+        return self.project
+
+    @property
+    def virtual_resource(self):
+        return self.resource
+
+    @property
+    def virtual_path(self):
+        return self
+
+    @cached_property
+    def canonical_project(self):
+        """Returns the unique project that matches the name that user has view access to.
+        If no match is found, returns None
+
+        Raises:
+            DuplicateProjectError - if project name is not unique on DX platform
+            NotFoundError - If project name doesn't exist on DNAnexus
+        """
+        if utils.is_valid_dxid(self.project, 'project'):
+            return self.project
+        else:
+            try:
+                proj_dict = dxpy.find_one_project(
+                    name=self.project, level='VIEW', zero_ok=True, more_ok=False)
+                if proj_dict is None:
+                    raise ProjectNotFoundError('No projects were found with given name ({})'
+                                               .format(self.project))
+                return proj_dict['id']
+            except DXError as e:
+                raise DuplicateProjectError('Duplicate projects were found with given name ({})'
```

this seems like a dangerous check, because it's possible you'd swallow up a different error (e.g., about authentication or something else). can you list projects with a limit of 2 and then handle 0/1/2 projects with appropriate exceptions?

---

Changed DXError to DXSearchError which is only raised when there are multiple projects.
Since find_one_project does not have a limit argument, the only other way is to pass zero_ok=False, in which case DXSearchError is raised for 0/2 projects and there's no way to differentiate the two scenarios.